### ISSUE TITLE
Add explicit documentation about C/C++ feature usage rule

### ISSUE THREAD

**Issue by huoyaoyuan** [dotnet/runtime#70239]
Versioning of C++ has been painful for a long time. A large amount of CLR code was written almost 20 years ago, before a usable C++ standard existed. During examination about modernizing the code base, I'm quite unsure for what C++ features we can depend on. And I don't have all the environment we target to.

The things I can find in the repo is the C 11/C++ 11 standard configuration, and some GCC/clang version detection. However, C++ standard has always been abstract that we need to consider every actual implementation we use. Moreover, a feature exists don't mean we should use it, with concerns about deployment or binary size.

With adding documentation, we should be able to use modern C/C++ features with more confident.

The documentation should contain the following points:

- The versions of compilers we use in CI, retail and daily workflow
This should be a hard limit. Any feature not supported by these versions should not be used.
This should especially focus on infamous platforms. For example, the [linux requirement](https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md) doc specifies clang9 which supports quite recent features, however [tizen build](https://github.com/dotnet/runtime/issues/13440) mentions clang5. I can't figure out what compiler FreeBSD build uses without inspecting the build scripts manually. This helps tracking which platforms are requiring old compilers.
- The versions of OS and distros we need intentionally support, and there libc version.
It can be found [here](https://github.com/dotnet/runtime/issues/13440), but not clear to every runtime contributor (I found it just when writing this issue).
- The components of C/C++ standard library we are allowed and not allowed to use.
Sometimes we couldn't accept the cost of binary size, for example some C++ STL. Sometimes the behavior varies between platforms, for example `atol` which depends on `sizeof(long)`. For some cases I can easily infer the reason based on our replacement implementation, but for many cases I have totally no idea.

The reason for asking for this documentation is I have got several questions related, for instance:
- We were using and mimicking MSVC `_finite` instead of C standard `isfinite`. The IEEE754 semantic is clear and should not differ between implementation, but the two functions are end up calling different underlying implementations, which are both available in `ucrtbase.dll`.
Is it safe to switch to the standard one since they are both available in UCRT? Are there any performance considerations? (Though I guess there's not, since CLR shouldn't be using FP a lot, and for managed side we have managed implementation.)
- Why don't we enable `STDCPP_COMPACT` for MSVC, and ends up using self-defined `type_traits` etc? With UCRT, the versioning concerns should be decreased for Windows, and we are requiring a very recent version of MSVC.
- There are a lot of different flavors of numbers defined and used in the codebase, except newly written JIT and hosting code. Can we convert them to `stdint` in a batch? Some of them (for example windows `ULONG` are really confusing.

**Comment by danmoseley** [comment]
@AaronRobinsonMSFT

**Comment by AaronRobinsonMSFT** [comment]
@huoyaoyuan This is a huge but welcome ask - thank you. This is an area I am passionate about and am endeavoring to mitigate. Official document on our path forward is a good idea and I will take this upon myself to provide a design document on how I see moving toward the eventual goal to modernize our use of C++ in the dotnet/runtime code base.

A few issues related to C++ and the runtime that are worth linking:
- https://github.com/dotnet/runtime/issues/66154
- https://github.com/dotnet/runtime/issues/37310
- https://github.com/dotnet/runtime/issues/67470
- https://github.com/dotnet/runtime/issues/4107

**Comment by huoyaoyuan** [comment]
I want to gather GCC/Clang/glibc requirements first, especially for Tizen/FreeBSD.

There are already some STL components used in newly written libs, and should be marked as safe first.

**Comment by huoyaoyuan** [comment]
I really want to use `std::filesystem`, but it comes from C++17.
If clang5 is no longer used, we can investigate and try it, at least in test suites.

**Comment by AaronRobinsonMSFT** [comment]
> I really want to use `std::filesystem`, but it comes from C++17. If clang5 is no longer used, we can investigate and try it, at least in test suites.

In the test suites this is a possibility. In the product, this is likely a non-starter - at least for quite a while. The biggest issue with many C++ APIs is exceptions. Many C++ APIs with through exceptions that are simply not expected in many code paths and can play havoc if they occur. This is something I and others would like to address but it is a very long term goal. In the near term the goal would be to simply be able to consume the official system headers instead of our mocked out versions.

**Comment by huoyaoyuan** [comment]
> The biggest issue with many C++ APIs is exceptions.

Yes, I realized this.
In newly written code we are defining real PAL abstracting Windows and Unix, instead of calling simulations "pal", there is real abstraction under `coreclr\minipal`. If we want to add more "real PAL", when would it be placed?

**Comment by huoyaoyuan** [comment]
We need also share some abstractions between corehost and coreclr, namely UTF-8/UTF-16 parity handling. Where would it be placed to?

**Comment by AaronRobinsonMSFT** [comment]
> We need also share some abstractions between corehost and coreclr, namely UTF-8/UTF-16 parity handling.

I don't know if that is true. The corehost default and remains in the host's default encoding UTF16 for Windows and UTF8 everywhere else. This dichotomy isn't the same in coreclr. Coreclr is presently UTF16-centric the majority of the time. I'd like to change this but that again isn't the same as corehost which is based on the host platform. There might be opportunities here but that discussion is a long way off I think.

**Comment by huoyaoyuan** [comment]
@AaronRobinsonMSFT is it safe to use STL containers like vector and string?

The basic dynamic sized containers have wide usage. Concerns I can see:

- OOM, while our custom implementations aren't handling at all.
- They shouldn't be throwing exceptions?
- Optimized storage. If we can bring some dependencies from boost, we can use `small_vector`
- Heap management. It should be not a problem if our custom implementations are already using standard `new`?

**Comment by AaronRobinsonMSFT** [comment]
> @AaronRobinsonMSFT is it safe to use STL containers like vector and string?

No it is not. We cannot currently permit exceptions coming from any STL collection.

> OOM, while our custom implementations aren't handling at all.
> They shouldn't be throwing exceptions?

Correct.

> Optimized storage. If we can bring some dependencies from boost, we can use small_vector

Boost is not going to be an option for this repo.

> Heap management. It should be not a problem if our custom implementations are already using standard new?

Can you elaborate on this? I don't understand the question/comment.

**Comment by huoyaoyuan** [comment]
> > Heap management. It should be not a problem if our custom implementations are already using standard new?
> 
> Can you elaborate on this? I don't understand the question/comment.

STL containers can be customized for allocators, so it shouldn't be a problem if we need one.

> We cannot currently permit exceptions coming from any STL collection.

Did a search for STL containers with noexcept. C++ is so complex. Ouch.

> Boost is not going to be an option for this repo.

I wonder the reason. Boost license looks similar to MIT, but does have difference.

**Comment by AaronRobinsonMSFT** [comment]
> STL containers can be customized for allocators, so it shouldn't be a problem if we need one.

Yes, that is true and something I think we should explore. However, there are very complex rules around STL allocators and they have severe limitations in practice. It isn't an impossible approach but not something we are anywhere close to being able to start considering.

> Did a search for STL containers with noexcept. C++ is so complex. Ouch.

Yes it is. C++ is complex because of the performance ethos it aligns with and the diverse platforms it is designed to support - all of them.

> I wonder the reason. Boost license looks similar to MIT, but does have difference.

This isn't a license perspective, but more akin to the opinionated perspective that Boost aligns with. It is a diverse and useful suite of APIs but doesn't have the rigor or governance that ISO-C++ has. Boost often has APIs that eventually get into the standard and that is a good thing, but it is only after a lot of time and consideration that that occurs. I value Boost, but the .NET runtime is not the place to consume non-standard libraries that don't adhere to the same requirements are the C++ standard.

**Comment by AaronRobinsonMSFT** [comment]
@huoyaoyuan You might want to checkout the [native containers](https://github.com/dotnet/runtime/tree/main/src/native/containers) that have been written in C. These are designed to be used in all runtime implementations (NativeAOT, CoreCLR, Mono) and are immediately consumable in CoreCLR in many cases. These aren't C++ for reasons around sharing and least-surprise, but one could see an argument for creating C++ wrappers around them that adhere to C++ idioms and allow us to use common APIs found in the [`algorithms` library](https://en.cppreference.com/w/cpp/algorithm).

Other more modern C++ libraries like `filesystem` are, as I've mentioned, a long way off. They impose certain semantics that don't always align with the .NET runtime codebase and this makes using them more effort and less benefit in the long run. Not something we shouldn't strive for, but I am dubious using them could ever be justified in the cost/benefit calculus.
